dask/dataframe/io/hdf.py

```diff
@@ -506,24 +410,141 @@ def read_hdf(
             "When assuming pre-partitioned data, data must be "
             "read in its entirety using the same chunksizes"
         )
-    from ..multi import concat
-
-    return concat(
-        [
-            _read_single_hdf(
-                path,
-                key,
-                start=start,
-                stop=stop,
-                columns=columns,
-                chunksize=chunksize,
-                sorted_index=sorted_index,
-                lock=lock,
-                mode=mode,
-            )
-            for path in paths
-        ]
+
+    # Build metadata
+    with pd.HDFStore(paths[0], mode=mode) as hdf:
+        meta_key = _expand_key(key, hdf)[0]
+    meta = pd.read_hdf(paths[0], meta_key, mode=mode, stop=0)
+    if columns is not None:
+        meta = meta[columns]
+
+    # Common kwargs
+    if meta.ndim == 1:
+        common_kwargs = {"name": meta.name, "mode": mode}
+    else:
+        common_kwargs = {"mode": mode}
+
+    # Build parts
+    parts, divisions = build_parts(
+        paths, key, start, stop, chunksize, sorted_index, mode
+    )
+
+    # Construct Layer and Collection
+    label = "read-hdf-"
+    name = label + tokenize(paths, key, start, stop, sorted_index, chunksize, mode)
+    if len(divisions) != len(parts) + 1:
+        divisions = [None] * (len(parts) + 1)
```

Comment:
This is odd to me. `parts` and `divisions` are both computed in the same place -- why not ensure they are the expected shape when doing so?

Reply:
Moved the null-divisions logic into `_build_parts`